mlflow/genai/judges/optimizers/dspy.py

```diff
@@ -0,0 +1,177 @@
+"""DSPy-based alignment optimizer implementation."""
+
+import logging
+from abc import abstractmethod
+from typing import Any
+
+from pydantic import PrivateAttr
+
+from mlflow.entities.trace import Trace
+from mlflow.exceptions import MlflowException
+from mlflow.genai.judges import make_judge
+from mlflow.genai.judges.base import AlignmentOptimizer, Judge
+from mlflow.genai.judges.optimizers.dspy_utils import (
+    agreement_metric,
+    create_dspy_signature,
+    trace_to_dspy_example,
+)
+from mlflow.genai.judges.utils import get_default_model
+from mlflow.utils.annotations import experimental
+
+logger = logging.getLogger(__name__)
+
+
+@experimental(version="3.4.0")
+class DSPyAlignmentOptimizer(AlignmentOptimizer):
+    """
+    Abstract base class for DSPy-based alignment optimizers.
+
+    Provides common functionality for converting MLflow traces to DSPy examples
+    and handling DSPy program compilation.
+    """
+
+    _logger: logging.Logger = PrivateAttr()
+    _model: str = PrivateAttr()
+
+    @property
+    def model(self) -> str:
+        """Get the model used by this optimizer."""
+        return self._model
+
+    def __init__(self, model: str | None = None, **kwargs):
+        """
+        Initialize DSPy optimizer with common parameters.
+
+        Args:
+            model: Model to use for DSPy optimization. If None, uses get_default_model().
+            **kwargs: Additional keyword arguments.
+        """
+        super().__init__(**kwargs)
+        # Initialize private variables using PrivateAttr
+        self._logger = logging.getLogger(self.__class__.__name__)
+        # Private member variable for the model, defaulted to get_default_model()
+        self._model = model if model is not None else get_default_model()
+
+    @abstractmethod
+    def _dspy_optimize(self, program, examples, metric_fn) -> Any:
+        """
+        Perform DSPy optimization with algorithm-specific parameters.
+
+        Each implementation can decide how to split the data internally if needed.
+
+        Args:
+            program: The DSPy program to optimize
+            examples: Examples for optimization (implementations decide how to split)
+            metric_fn: Metric function for optimization
+
+        Returns:
+            Optimized DSPy program
+        """
+
+    def _lower_to_dspy(self, judge: Judge) -> Any:
+        """Lowers the judge into a Predict dspy module"""
+        try:
+            import dspy
+
+            class CustomPredict(dspy.Predict):
+                """Custom DSPy Predict class that allows passing an LM to the forward method."""
+
+                def __init__(self, judge):
+                    super().__init__(create_dspy_signature(judge))
+                    self._lm = dspy.LM(model=judge.model)
+
+                def forward(self, *args, **kwargs):
+                    # If an LM is supplied via kwargs, use that, else use self.lm
+                    lm = kwargs.pop("lm", self._lm)
```

Comment:
Is this "override the LM" capability necessary? If so, can we supplement the inline comment to explain why? If not, can we remove it to reduce the potential for accidental / unintentional changes in the judge model?

Reply:
It is necessary so that the DSPy program uses the same model as the judge rather than the model used during optimization. Some DSPy modules actually support this (e.g., `ChainOfThought`) but not `Predict` which we use here, hence the wrapper. I 'll add to the comments to explain this.